mesh/mesh.go

```diff
@@ -181,10 +187,50 @@ func (m *Mesh) ValidateLayer(lyr *types.Layer) {
 			m.With().Error("failed to push transactions", log.Err(err))
 			break
 		}
+		m.logStateRoot(layerId)
+		m.setLayerHash(layerId)
 	}
+	m.persistLayerHash()
 	m.Info("done validating layer %v", lyr.Index())
 }
 
+func (m *Mesh) logStateRoot(layerId types.LayerID) {
+	m.Event().Info("end of layer state root",
+		log.LayerId(layerId.Uint64()),
+		log.String("state_root", util.Bytes2Hex(m.TxProcessor.GetStateRoot().Bytes())),
+	)
+}
+
+func (m *Mesh) setLayerHash(layerId types.LayerID) {
+	validBlockIds, err := m.ContextuallyValidBlock(layerId)
```

Comment:
this is code duplication from `ExtractUniqueOrderedTransactions` , try to refactor